### ISSUE TITLE
[FEATURE]: Enable no resampling by raster tiler

### ISSUE THREAD

**Issue by rbavery** [CosmiQ/solaris#290]
_Thank you for helping us improve `solaris`!_

## Summary of the bug
I ran the rio tiler tutorial example on a Landsat Analysis Ready Data Scene to tile it into 512x512 chips. But new negative values are added that weren't there before in the original dataset (which only had -9999).

## Steps to reproduce the bug
Please either paste sample code used to generate the buggy behavior below, or provide step-by-step instructions to reproduce the problem.

```
import solaris as sol
import os

raster_tiler = sol.tile.raster_tile.RasterTiler(dest_dir='test_solaris',  # the directory to save images to
                                                src_tile_size=(512, 512),  # the size of the output chips
                                                verbose=True,
                                               nodata=-9999)
raster_tiler.tile('/mnt/cropmaskperm/test-landsat/scene/LT05_L1TP_029031_20050316_20160912_01_T1_sr.tif')
bigarr = skio.imread('/mnt/cropmaskperm/test-landsat/scene/LT05_L1TP_029031_20050316_20160912_01_T1_sr.tif')
arr = skio.imread("test_solaris/LT05_L1TP_029031_20050316_20160912_01_T1_sr_696765_4625505.tif")
np.unique(arr[...,2])
np.unique(bigarr)
8578 in np.unique(bigarr)
```
```
array([-9999, -9998, -9997, ...,  8467,  8578, 19423], dtype=int16) #chip arr
array([-9999,  -898,  -897, ...,  8923,  8960, 20000], dtype=int16) # src arr
False # negative and positive values are changed, the src arr value should be in the chip arr
```
Run the rio tiler on a Landsat scene and check if the chip array values are in the source scene. If not it means values have been mutated in the tiling step.

## Expected behavior
The tiling step shouldn't mutate the values 

## Environment information
- OS: ubuntu 16
- `solaris` version: 0.1.3
- python version: 3.6

**Comment by rbavery** [comment]
I missed that rasterio's read function probably resamples the input if it is not a regular read of a whole image. The description leaves it unclear if the default nearest neighbor resampling is used if no resampling method is specified but window is specified, do you know what the behavior is when reading by window (as in the raster tiler source)?

From https://rasterio.readthedocs.io/en/stable/api/rasterio.io.html
```
resampling (Resampling) – By default, pixel values are read raw or interpolated using a nearest neighbor algorithm from the band cache. Other resampling algorithms may be specified. Resampled pixels are not cached.
```

I think for chipping rasters where the source and the chips are the same projection, shouldn't it be possible to not resample at all and pad edges with nodata?

**Comment by nrweir** [comment]
Hi @rbavery,

As it's currently implemented, `solaris` automatically re-projects to the appropriate UTM zone if you don't provide the `dest_crs` argument when initializing the `RasterTiler`. Perhaps this is either misguided or should be made more explicitly clear (with a warning or something like that when reprojecting) so it doesn't silently re-project imagery without users knowing. The logic in implementing it this way was that many computer vision folks might not be super familiar with CRSs, and we tried to funnel everything to a fairly standard CRS for the sake of a simple user experience; however, we clearly need to also ensure that experts are aware of what's going on under the hood.

If you provide the EPSG code for your source CRS as a `dest_crs` argument, do you still get the pixel value changes?

Thank you for your help making `solaris` a more useful library!

**Comment by rbavery** [comment]
Thanks for the quick reply! Providing a dest_crs does indeed keep the original values. One issue with the requirement of providing a dest_crs to stop UTM reprojection is that it requires your src projection to have an EPSG code, which the Landsat ARD product and other products do not have. This is the Proj4 string for Landsat ARD, I haven't been able to find an EPSG code that matches:

` '+proj=aea +lat_0=23 +lon_0=-96 +lat_1=29.5 +lat_2=45.5 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs=True'`

I'm working on my fork of solaris to fix this issue for myself but I'm interested in merging changes back in if they look useful for others. 

On the utm reprojecting, I've found the part where the reprojection occurs:

```
            if self.proj_unit not in ['meter', 'metre']:
                if self.verbose:
                    print("Input CRS is not metric. "
                          "Reprojecting the input to UTM.")
                self.src = reproject(self.src,
                                     resampling_method=self.resampling,
                                     dest_path=os.path.join(self.dest_dir,
                                                            'tmp.tif'))
```

I think the reprojection should occur only if the user specifies that they want the units to be in meters, but I think providing a warning is good too. Would you support this if statement instead checking for an argument `project_to_meters` and reprojecting if this is set to True (I would think it should be False default but others might prefer otherwise).

**Comment by nrweir** [comment]
Hi @rbavery,

Great point on the EPSG code issue. Let me chat with the team here regarding defaulting to re-project or leaving it alone. This topic came up a lot as we were writing the initial implementation and we eventually settled on the solution that's implemented, but given this context might be willing to switch.

**Comment by rbavery** [comment]
I see that this core function returns epsg codes for use in reprojecting in the tilers

```

def _check_crs(input_crs):
    """Convert CRS to the integer format passed by ``solaris``."""
    if isinstance(input_crs, dict):
        # assume it's an {'init': 'epsgxxxx'} dict
        out_crs = int(input_crs['init'].lower().strip('epsg:'))
    elif isinstance(input_crs, str):
        # handle PROJ4 strings, epsg strings, wkt strings
        out_crs = rasterio.crs.CRS.from_string(input_crs).to_epsg()
    elif isinstance(input_crs, rasterio.crs.CRS):
        out_crs = input_crs.to_epsg()
    elif isinstance(input_crs, int):
        out_crs = input_crs
    elif input_crs is None:
        out_crs = input_crs
    return out_crs
```

but the tilers convert these epsg codes to rasterio CRS objects or use the CRS object that is an attribute of the DatasetReader object (src).

```
dst_transform, width, height = calculate_default_transform(
                        self.src.crs, CRS.from_epsg(self.dest_crs),
                        self.src.width, self.src.height, *tb,
                        dst_height=self.dest_tile_size[0],
                        dst_width=self.dest_tile_size[1])
```
Could we instead return a rasterio CRS object from `_check_crs` and rewrite it to catch proj4 strings like this? Then the tilers could use these CRS objects directly rather than converting the EPSG code to proj4 within the tiler functions. This would allow support for proj4 strings. I'm working on this on my fork and am happy to merge it back if this sounds beneficial.

```
def _check_crs(input_crs):
    """Convert CRS to the integer format passed by ``solaris``."""
    if isinstance(input_crs, dict):
        # assume it's an {'init': 'epsgxxxx'} dict
        out_crs = int(input_crs['init'].lower().strip('epsg:'))
        out_crs = rasterio.crs.CRS.from_epsg(out_crs)
    elif isinstance(input_crs, str):
        # handle PROJ4 strings, epsg strings, wkt strings
        out_crs = rasterio.crs.CRS.from_proj4(input_crs)
    elif isinstance(input_crs, rasterio.crs.CRS):
        out_crs = input_crs
    elif isinstance(input_crs, int):
        out_crs = rasterio.crs.CRS.from_epsg(input_crs)
    elif input_crs is None:
        out_crs = input_crs
    return out_crs
```

**Comment by nrweir** [comment]
Yeah, that works for me. We were trying to keep things as simple as possible by just passing around EPSG codes, but wasn't considering the CRSs that don't have one. There are unfortunately a number of places where `_check_crs` and EPSG code CRSs are used, so this might require a fair amount of finagling to debug everything. However, if you want to fix it, go for it and we'll be very grateful!

**Comment by nrweir** [comment]
@rbavery - consensus on this side is that keeping source CRS rather than defaulting to UTM reprojection is a better plan. If you have time to implement, great; if not, we'll try to squeeze it in before 0.2.0.

**Comment by rbavery** [comment]
Great! I think I have a working implementation now for the tilers but need to fix some other tests, I'll make a WIP PR today.

**Comment by nrweir** [comment]
Resolved with #294!
